### PR TITLE
Add PCA GPU backend in oneAPI interfaces

### PR DESCRIPTION
**Pull request by rlnx**
- Enable DPC++ build via Bazel
- Enabled DPC++ tests in Bazel
- Added PCA tests
- Test system prototype
- Abstractions for linear algebra

### CODE REVIEW

**Comment by rlnx** [comment]
@michael-smirnov, please take a look at `pca/tests` and `backend/linalg`, these are the main changes

File: cpp/oneapi/dal/algo/linear_kernel/BUILD [hunk 1]
```diff
@@ -17,17 +17,33 @@ dal_module(
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/linear_kernel/BUILD:17]: Is this file actually related to PCA?

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc_test.cpp [hunk 2]
```diff
@@ -17,8 +17,7 @@
 #include <CL/sycl.hpp>
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc_test.cpp:17]: Is this file actually related to PCA?

File: cpp/oneapi/dal/algo/kmeans_init/BUILD [hunk 3]
```diff
@@ -17,17 +17,33 @@ dal_module(
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/kmeans_init/BUILD:17]: Is this file actually related to PCA?

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc_test.cpp [hunk 4]
```diff
@@ -17,8 +17,8 @@
 #include <CL/sycl.hpp>
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc_test.cpp:17]: Is this file actually related to PCA?

File: cpp/oneapi/dal/algo/kmeans/BUILD [hunk 5]
```diff
@@ -17,17 +17,33 @@ dal_module(
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/kmeans/BUILD:17]: It's not related to PCA anyhow. Please check branches.

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/infer_train_reg_kernel_dpc_test.cpp [hunk 6]
```diff
@@ -17,8 +17,8 @@
 #include <CL/sycl.hpp>
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/infer_train_reg_kernel_dpc_test.cpp:17]: It's not about PCA

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/infer_train_cls_kernel_dpc_test.cpp [hunk 7]
```diff
@@ -17,8 +17,8 @@
 #include <CL/sycl.hpp>
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/infer_train_cls_kernel_dpc_test.cpp:17]: It's not PCA related file

File: cpp/oneapi/dal/algo/decision_forest/BUILD [hunk 8]
```diff
@@ -16,30 +16,62 @@ dal_module(
     ]
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/decision_forest/BUILD:16]: It's not PCA related file. Can this changes be separated into another PR?

File: cpp/oneapi/dal/algo/rbf_kernel/BUILD [hunk 9]
```diff
@@ -17,17 +17,33 @@ dal_module(
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/rbf_kernel/BUILD:17]: It's not PCA related cahnge

File: cpp/oneapi/dal/algo/pca/train_types.hpp [hunk 10]
```diff
@@ -68,16 +69,22 @@ class ONEAPI_DAL_EXPORT train_result {
         return *this;
     }
 
-    auto& set_explained_variance(const table& value) {
-        set_explained_variance_impl(value);
+    auto& set_variances(const table& value) {
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/pca/train_types.hpp:72]: Shouldn't it be marked as `inline`?

File: cpp/oneapi/dal/algo/pca/backend/cpu/train_kernel_cov.cpp [hunk 11]
```diff
@@ -46,8 +46,8 @@ static train_result<Task> call_daal_kernel(const context_cpu& ctx,
     auto arr_data = row_accessor<const Float>{ data }.pull();
     auto arr_eigvec = array<Float>::empty(column_count * component_count);
     auto arr_eigval = array<Float>::empty(1 * component_count);
-    auto arr_means = array<Float>::empty(1 * component_count);
-    auto arr_vars = array<Float>::empty(1 * component_count);
+    auto arr_means = array<Float>::empty(1 * column_count);
+    auto arr_vars = array<Float>::empty(1 * column_count);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/pca/backend/cpu/train_kernel_cov.cpp:50]: Should we actually multiply it? As for me it doesn't provide any additional explicitness.
```suggestion
    auto arr_means = array<Float>::empty(column_count);
    auto arr_vars = array<Float>::empty(column_count);
```

File: cpp/oneapi/dal/detail/common.hpp [hunk 12]
```diff
@@ -72,4 +72,76 @@ Object make_from_pointer(typename Object::pimpl::element_type* pointer) {
...
+    else if (t == data_type::int64) {
+        return sizeof(int64_t);
+    }
+    else if (t == data_type::uint32) {
+        return sizeof(uint32_t);
+    }
+    else if (t == data_type::uint64) {
+        return sizeof(uint64_t);
+    }
+    return 0;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/detail/common.hpp:94]: Shouldn't it cause compile or runtime error?

File: cpp/oneapi/dal/detail/common.hpp [hunk 13]
```diff
@@ -72,4 +72,76 @@ Object make_from_pointer(typename Object::pimpl::element_type* pointer) {
     return pimpl_accessor().template make_from_pointer<Object>(pointer);
 }
 
+constexpr std::int64_t get_data_type_size(data_type t) {
+    if (t == data_type::float32) {
+        return sizeof(float);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/detail/common.hpp:77]: Shouldn't this `if`s be marked as `constexpr`? at least for consistency with `make_data_type_impl` function?

File: cpp/oneapi/dal/table/detail/common.hpp [hunk 14]
```diff
@@ -20,78 +20,4 @@
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/table/detail/common.hpp:20]: What is the purpose of this file?

File: cpp/oneapi/dal/test/macro.hpp [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+*******************************************************************************/
+
+#pragma once
+
+// Disable clang-format as it dramatically
+// affects redability of macro definitions
+// clang-format off
+
+#define _TS_STRINGIFY(x) #x
+#define _TS_EXPAND(...)  __VA_ARGS__
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/test/macro.hpp:24]: Usually all variables/functions/macros/etc that starts with underscore refers to internal implementation. Why there is no "public" macros in this file?

File: cpp/oneapi/dal/test/macro.hpp [hunk 16]
```diff
@@ -0,0 +1,114 @@
+/*******************************************************************************
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/test/macro.hpp:1]: Do we really need this macros? Is there a way to do this using templates?

**Review by KulikovNikita — COMMENTED** [review]
Please check if this PR contains commits from another ones?

File: cpp/oneapi/dal/algo/kmeans_init/BUILD (hunk 3 above)
Comment (rlnx) [reply cpp/oneapi/dal/algo/kmeans_init/BUILD:17]: I've added support for DPC++ tests in Bazel to be able to run DPC++ PCA tests and had to fix existing rules/tests.

File: cpp/oneapi/dal/algo/kmeans/BUILD (hunk 5 above)
Comment (rlnx) [reply cpp/oneapi/dal/algo/kmeans/BUILD:17]: I've added support for DPC++ tests in Bazel to be able to run DPC++ PCA tests and had to fix existing rules/tests.

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/infer_train_reg_kernel_dpc_test.cpp (hunk 6 above)
Comment (rlnx) [reply cpp/oneapi/dal/algo/decision_forest/backend/gpu/infer_train_reg_kernel_dpc_test.cpp:17]: I've added support for DPC++ tests in Bazel to be able to run DPC++ PCA tests and had to fix existing rules/tests.

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/infer_train_cls_kernel_dpc_test.cpp (hunk 7 above)
Comment (rlnx) [reply cpp/oneapi/dal/algo/decision_forest/backend/gpu/infer_train_cls_kernel_dpc_test.cpp:17]: I've added support for DPC++ tests in Bazel to be able to run DPC++ PCA tests and had to fix existing rules/tests.

File: cpp/oneapi/dal/test/macro.hpp (hunk 15 above)
Comment (rlnx) [reply cpp/oneapi/dal/test/macro.hpp:24]: I don't want this macros to be used explicitly in tests implementation. Their use should be limited by other more problem-specific macros, like `ALGO_TEST_CASE` (see `dal/test/common.hpp`) required to organize Catch2 test cases enumeration.

File: cpp/oneapi/dal/algo/kmeans_init/BUILD (hunk 3 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/kmeans_init/BUILD:17]: Can this support be implemented for the only algorithm i.e. PCA?

File: cpp/oneapi/dal/test/macro.hpp (hunk 15 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/test/macro.hpp:24]: Thanx for explanation!

File: cpp/oneapi/dal/test/macro.hpp (hunk 16 above)
Comment (rlnx) [reply cpp/oneapi/dal/test/macro.hpp:1]: No templates don't help here, generally macros are needed to simplify declaration of tests cases for algorithms, to iterate over all possible combinations of Floats and Methods, see `ALGO_TEST_CASE` (`dal/test/common.hpp`).

```c++
ALGO_TEST_CASE("test case name", (float, double), (method::m1, method::m2, method::m3)) {
   // This test case will run 2 x 3 = 6 times with the different Float and Method template parameters
   // For more details, have a look at Catch2 type parametrised test cases, this one is wrapper over `TEMPLATE_TEST_CASE`
}
```

Comment (rlnx) [reply cpp/oneapi/dal/test/macro.hpp:1]: In other words, those macros are needed for template instantiation, you cannot **explicitly** instantiate template using the other template.

Comment (KulikovNikita) [reply cpp/oneapi/dal/test/macro.hpp:1]: Yeap. It was out of my scope. Thank you

File: cpp/oneapi/dal/table/detail/common.hpp (hunk 14 above)
Comment (rlnx) [reply cpp/oneapi/dal/table/detail/common.hpp:20]: It's expected to contain common definitions for tables. I assume this will be needed  a bit later, there is a lot of code that can be generalized and reused for tables and can be placed here.

File: cpp/oneapi/dal/detail/common.hpp (hunk 12 above)
Comment (rlnx) [reply cpp/oneapi/dal/detail/common.hpp:94]: it should.

File: cpp/oneapi/dal/detail/common.hpp (hunk 13 above)
Comment (rlnx) [reply cpp/oneapi/dal/detail/common.hpp:77]: It cannot be marked as `constexpr` since the value of `t` is determined at runtime. But this function can be evaluated at compile time if `t` is known at compile time, so the function itself marked as `constexpr`.

File: cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc_test.cpp (hunk 4 above)
Comment (rlnx) [reply cpp/oneapi/dal/algo/kmeans/backend/gpu/train_kernel_lloyd_dense_dpc_test.cpp:17]: I've added support for DPC++ tests in Bazel to be able to run DPC++ PCA tests and had to fix existing rules/tests.

File: cpp/oneapi/dal/algo/decision_forest/BUILD (hunk 8 above)
Comment (rlnx) [reply cpp/oneapi/dal/algo/decision_forest/BUILD:16]: I've added support for DPC++ tests in Bazel to be able to run DPC++ PCA tests and had to fix existing rules/tests.

File: cpp/oneapi/dal/algo/rbf_kernel/BUILD (hunk 9 above)
Comment (rlnx) [reply cpp/oneapi/dal/algo/rbf_kernel/BUILD:17]: I've added support for DPC++ tests in Bazel to be able to run DPC++ PCA tests and had to fix existing rules/tests.

File: cpp/oneapi/dal/algo/pca/train_types.hpp (hunk 10 above)
Comment (rlnx) [reply cpp/oneapi/dal/algo/pca/train_types.hpp:72]: No, this is class member function, it's `inline` by design

File: cpp/oneapi/dal/algo/pca/backend/cpu/train_kernel_cov.cpp (hunk 11 above)
Comment (rlnx) [reply cpp/oneapi/dal/algo/pca/backend/cpu/train_kernel_cov.cpp:50]: This helps to understand actual shape of the arr_vars, because it's used for `homogen_table` construction

File: cpp/oneapi/dal/algo/kmeans_init/BUILD (hunk 3 above)
Comment (rlnx) [reply cpp/oneapi/dal/algo/kmeans_init/BUILD:17]: No, I need all tests to be passed

File: cpp/oneapi/dal/algo/linear_kernel/BUILD (hunk 1 above)
Comment (rlnx) [reply cpp/oneapi/dal/algo/linear_kernel/BUILD:17]: I've added support for DPC++ tests in Bazel to be able to run DPC++ PCA tests and had to fix existing rules/tests.

File: cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc_test.cpp (hunk 2 above)
Comment (rlnx) [reply cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_dense_dpc_test.cpp:17]: I've added support for DPC++ tests in Bazel to be able to run DPC++ PCA tests and had to fix existing rules/tests.

**Review by KulikovNikita — APPROVED** [review]

**Comment by napetrov** [comment]
@Mergifyio backport rls/onedal-beta09-rls
